map2loop/data_checks.py

```diff
@@ -0,0 +1,882 @@
+        for key in section_dict.keys():
+            if key not in allowed_keys:
+                logger.error(f"Key '{key}' is not an allowed key in the '{section_name}' section.")
+                raise ValueError(f"Key '{key}' is not an allowed key in the '{section_name}' section.")
+    
+    # 4) check if minimum fault length is a number
+    mfl = config_dict.get("fault", {}).get("minimum_fault_length", None)
+    if mfl is not None and not isinstance(mfl, (int, float)):
+        logger.error("minimum_fault_length must be a number.")
+        raise ValueError(f"minimum_fault_length must be a number, instead got: {type(mfl)}")
+    
+@beartype.beartype
+def validate_geometry(
+    geodata: geopandas.GeoDataFrame,
+    expected_geom_types: List[type],
+    datatype_name: str
+) -> Tuple[bool, str]:
+
+    # 1. Check if all geometries are valid
+    if not geodata.geometry.is_valid.all():
+        logger.error(f"Invalid geometries found in datatype {datatype_name}. Please fix them before proceeding.")
+        return True, f"Invalid geometries found in datatype {datatype_name}"
+    
+    # 2. Check if all geometries are of the expected types
+    if not geodata.geometry.apply(lambda geom: isinstance(geom, tuple(expected_geom_types))).all():
+        invalid_types = geodata[~geodata.geometry.apply(lambda geom: isinstance(geom, tuple(expected_geom_types)))]
+        invalid_indices = invalid_types.index.tolist()
+        expected_types_names = ', '.join([geom_type.__name__ for geom_type in expected_geom_types])
+        logger.error(
+            f"Datatype {datatype_name}: Invalid geometry types found. Expected types: {expected_types_names}. "
+            f"Rows with invalid types: {invalid_indices}"
+        )
+        return True, (
+            f"Invalid geometry types found in datatype {datatype_name}. "
+            f"All geometries must be {expected_types_names}."
+        )
+    
+    # If all checks pass
+    logger.debug(f"Geometry validation passed for datatype {datatype_name}")
+    return False, ""
+
+
+@beartype.beartype
+def validate_id_column(
+    geodata: geopandas.GeoDataFrame,
+    config: dict,
+    id_config_key: str, 
+    geodata_name: str
+) -> Tuple[bool, str]:
+
+    # Retrieve the ID column name from the configuration
+    id_column = config.get(id_config_key)
+    
+    if not id_column:
+        error_msg = f"Configuration key '{id_config_key}' is missing."
+        logger.error(error_msg)
+        return (True, error_msg)
+    
+    if id_column in geodata.columns:
+        geodata[id_column] = pandas.to_numeric(geodata[id_column], errors='coerce')
+    
+        # Check for non-numeric values (which are now NaN after coercion)
+        if geodata[id_column].isnull().any():
+            error_msg = (
+                f"Datatype {geodata_name}: Column '{id_column}' "
+                f"(config key: '{id_config_key}') contains non-numeric or NaN values. "
```

Comment:
This is good, we could make the error message even better by providing the strings (or some of) for the values that couldn't be coerced